basics_of_logging_in_sst-core/Example02/Logger.h

```diff
@@ -0,0 +1,76 @@
+#ifndef LOGGER_H
+#define LOGGER_H
+
+#include <sst/core/output.h>
+#include <stdio.h>
+
+namespace Logger
+{
+    // Define the log levels.  Each level displays its own message and the
+    // levels above it.
+    //
+    enum LogLevel {
+        Fatal = 0,
+        Error = 100,
+        Warn  = 200,
+        Info  = 300,
+        Debug = 400,
+        Trace = 500
+    };
+
+    template<typename... Args> 
```

Comment:
This is an interesting approach, but the templating can lead to severe code bloat since you get a new version of the code for every combination of variables passed into the function.  This can lead to slow code due to more misses in the instruction cache.  If you really want this type of logger, we can add a version of all of Output's functions that takes a va_list argument.  That way you can get rid of the templating.  I would not encourage templating in this case.

Reply:
It is also possible to add an attribute so the compiler will warn the users when the types in the format string don't match the types of the corresponding parameters that are passed in.  This is done with:

__attribute__((format(printf, 6, 7)));

Not sure it this will work with the templated functions, but would for sure work with non-templated functions that called a va_list version of the calls.